File: src/omnia/data/structure/observable/ObservableGraph.java
package omnia.data.structure.observable;

import io.reactivex.rxjava3.core.Observable;
import omnia.data.structure.Graph;
import omnia.data.structure.Set;
import omnia.data.structure.tuple.Couplet;

public interface ObservableGraph<E> extends Graph<E>, ObservableDataStructure {

  @Override
  ObservableChannels<E> observe();

  interface ObservableChannels<E> extends ObservableDataStructure.ObservableChannels {

    @Override
    Observable<? extends Graph<E>> states();

    @Override
    Observable<? extends ObservableGraph.MutationEvent<E>> mutations();
  }

  interface MutationEvent<E> extends ObservableDataStructure.MutationEvent {

    @Override
    Graph<E> state();

    @Override
    Set<? extends ObservableGraph.GraphOperation<E>> operations();
  }

  interface GraphOperation<E> {

    static <E> Observable<AddNodeToGraph<E>> justAddNodeToGraphOperations(
        Observable<? extends GraphOperation<E>> flowable) {
      return flowable.flatMap(
          mutation -> mutation instanceof AddNodeToGraph<?>
              ? Observable.just((AddNodeToGraph<E>) mutation)
              : Observable.empty());
    }

    static <E> Observable<RemoveNodeFromGraph<E>> justRemoveNodeFromGraphOperations(
        Observable<? extends GraphOperation<E>> flowable) {
      return flowable.flatMap(
          mutation -> mutation instanceof RemoveNodeFromGraph<?>
              ? Observable.just((RemoveNodeFromGraph<E>) mutation)
              : Observable.empty());
    }

    static <E> Observable<AddEdgeToGraph<E>> justAddEdgeToGraphOperations(
        Observable<? extends GraphOperation<E>> flowable) {
      return flowable.flatMap(
          mutation -> mutation instanceof AddEdgeToGraph<?>
              ? Observable.just((AddEdgeToGraph<E>) mutation)
              : Observable.empty());
    }

    static <E> Observable<RemoveEdgeFromGraph<E>> justRemoveEdgeFromGraphOperations(
        Observable<? extends GraphOperation<E>> flowable) {
      return flowable.flatMap(
          mutation -> mutation instanceof RemoveEdgeFromGraph<?>
              ? Observable.just((RemoveEdgeFromGraph<E>) mutation)
              : Observable.empty());
    }
  }

  interface NodeOperation<E> extends GraphOperation<E> {
    E item();
  }

  interface EdgeOperation<E> extends GraphOperation<E> {
    Couplet<E> endpoints();
  }

  interface AddNodeToGraph<E> extends NodeOperation<E> {}

  interface RemoveNodeFromGraph<E> extends NodeOperation<E> {}

  interface AddEdgeToGraph<E> extends EdgeOperation<E> {}

  interface RemoveEdgeFromGraph<E> extends EdgeOperation<E> {}
}

File: src/omnia/data/structure/observable/ObservableGraph.kt
package omnia.data.structure.observable

import com.badoo.reaktive.observable.Observable
import com.badoo.reaktive.observable.mapNotNull
import omnia.data.structure.Graph
import omnia.data.structure.Set
import omnia.data.structure.tuple.Couplet

interface ObservableGraph<E : Any> : Graph<E>, ObservableDataStructure {

  override fun observe(): ObservableChannels<E>

  interface ObservableChannels<E : Any> : ObservableDataStructure.ObservableChannels {

    override fun states(): Observable<Graph<E>>

    override fun mutations(): Observable<MutationEvent<E>>
  }

  interface MutationEvent<E : Any> : ObservableDataStructure.MutationEvent {

    override fun state(): Graph<E>

    override fun operations(): Set<out GraphOperation<E>>
  }

  interface GraphOperation<E : Any> {
    companion object {

      fun <E : Any> justAddNodeToGraphOperations(operations: Observable<GraphOperation<E>>):
          Observable<AddNodeToGraph<E>> {
        return operations.mapNotNull { it as? AddNodeToGraph<E> }
      }

      fun <E : Any> justRemoveNodeFromGraphOperations(operations: Observable<GraphOperation<E>>):
          Observable<RemoveNodeFromGraph<E>> {
        return operations.mapNotNull { it as? RemoveNodeFromGraph<E> }
      }

      fun <E : Any> justAddEdgeToGraphOperations(operations: Observable<GraphOperation<E>>):
          Observable<AddEdgeToGraph<E>> {
        return operations.mapNotNull { it as? AddEdgeToGraph<E> }
      }

      fun <E : Any> justRemoveEdgeFromGraphOperations(operations: Observable<GraphOperation<E>>):
          Observable<RemoveEdgeFromGraph<E>> {
        return operations.mapNotNull { it as? RemoveEdgeFromGraph<E> }
      }
    }
  }

  interface NodeOperation<E : Any> : GraphOperation<E> {

    fun item(): E
  }

  interface EdgeOperation<E : Any> : GraphOperation<E> {

    fun endpoints(): Couplet<E>
  }

  interface AddNodeToGraph<E : Any> : NodeOperation<E>

  interface RemoveNodeFromGraph<E : Any> : NodeOperation<E>

  interface AddEdgeToGraph<E : Any> : EdgeOperation<E>

  interface RemoveEdgeFromGraph<E : Any> : EdgeOperation<E>
}